File: cf_worker/src/utils/text.ts
import type { NewsItem } from "../services/news";
import type { Labels } from "./i18n";

export function escapeHtml(input: string): string {
  return input
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;")
    .replace(/"/g, "&quot;")
    .replace(/'/g, "&#39;");
}

function stripCodeBlocks(text: string): string {
  return text.replace(/```[\s\S]*?```/g, "");
}

function sanitizeText(text: string, maxLen: number): string {
  const cleaned = stripCodeBlocks(text).replace(/\s+/g, " ").trim();
  if (cleaned.length > maxLen) {
    return cleaned.slice(0, maxLen).trim();
  }
  return cleaned;
}

export function parseActionItems(raw: string | null | undefined): string[] {
  if (!raw) {
    return [];
  }
  try {
    const parsed = JSON.parse(raw);
    if (Array.isArray(parsed)) {
      return parsed.map((item) => String(item));
    }
  } catch (_) {
    return [];
  }
  return [];
}

export function formatItemMessage(
  item: NewsItem,
  labels: Labels,
  overrides?: { title?: string; rationale?: string; actions?: string[] }
): string {
  const title = escapeHtml(sanitizeText(overrides?.title ?? item.title ?? labels.noTitle, 140));
  const rationale = escapeHtml(
    sanitizeText(overrides?.rationale ?? item.impact_rationale ?? labels.noRationale, 360)
  );
  const rawScore = item.impact_score ?? 1;
  const impactScore = Math.min(5, Math.max(1, Math.round(rawScore)));
  const actions = (overrides?.actions ?? parseActionItems(item.action_items_json))
    .map((action) => sanitizeText(action, 140))
    .filter((action) => action.length > 0)
    .slice(0, 4);
  const actionsBlock = actions.length
    ? actions.map((action) => `• ${escapeHtml(action)}`).join("\n")
    : `• ${labels.noActions}`;

  return [
    `<b>${title}</b>`,
    `<b>${labels.impact}:</b> ${impactScore}/5`,
    `<b>${labels.why}:</b> ${rationale}`,
    `<b>${labels.actions}:</b>`,
    actionsBlock,
    `<a href="${escapeHtml(item.url ?? "")}">${labels.source}</a>`
  ].join("\n");
}

File: bot/utils/text.py
import json
import html
from typing import List, Any, Dict


MAX_ACTION_ITEMS = 5


def parse_json_list(value: Any) -> List[str]:
    if value is None:
        return []
    if isinstance(value, list):
        return [str(x).strip() for x in value if str(x).strip()]
    if isinstance(value, str):
        try:
            data = json.loads(value)
            if isinstance(data, list):
                return [str(x).strip() for x in data if str(x).strip()]
        except json.JSONDecodeError:
            return [value.strip()] if value.strip() else []
    return [str(value).strip()]


def format_news_item(item: Dict[str, Any]) -> str:
    title = html.escape(item.get("title") or "Untitled")
    url = html.escape(item.get("url") or "")
    score = item.get("impact_score")
    rationale = html.escape(item.get("impact_rationale") or "")

    action_items = parse_json_list(item.get("action_items_json"))[:MAX_ACTION_ITEMS]
    if action_items:
        actions = "\n".join(f"- {html.escape(a)}" for a in action_items)
    else:
        actions = "- No action items provided"

    parts = [
        f"<b>{title}</b>",
        f"<a href=\"{url}\">Source</a>" if url else "",
        f"Impact score: {score}" if score is not None else "Impact score: n/a",
        f"Rationale: {rationale}" if rationale else "Rationale: n/a",
        "Action items:\n" + actions,
    ]
    return "\n".join(p for p in parts if p)
